File: CSharp/PinetreeChat.Domain/Services/UserService.cs
using PinetreeChat.Domain.Services.Exceptions;
using PinetreeChat.Entities;
using PinetreeChat.Interfaces.Repositories;
using System.Linq;

namespace PinetreeChat.Domain.Services
{
    public class UserService
    {
        private IUserRepository _userRepo;

        public UserService(IUserRepository userRepo)
        {
            _userRepo = userRepo;
        }        

        public void LogIn(string username)
        {
            if(username.Length > 20 || string.IsNullOrWhiteSpace(username))
            {
                throw new UsernameInvalidException(username);
            }

            var user = _userRepo.GetUsers().SingleOrDefault(u => u.Username.ToLower() == username.ToLower());
            if(user != null && user.IsLoggedIn)
            {
                throw new UserExistsException(username);
            }
            else if(user != null && !user.IsLoggedIn)
            {
                _userRepo.SetLoggedIn(user);
            }
            else
            {
                user = new User(username);
                _userRepo.AddUser(user);
                _userRepo.SetLoggedIn(user);
            }
        }

        public void LogOut(string username)
        {
            var user = _userRepo.GetUsers().SingleOrDefault(u => u.Username.ToLower() == username.ToLower());
            if (user == null)
            {
                return;
            }
            
            _userRepo.SetLoggedOut(user);
        }
    }
}

File: FSharp/PinetreeChat.Domain/UserService.fs
namespace PinetreeChat.Domain

module UserService = 
    open PinetreeChat.Entities
    open PinetreeChat.Domain.Validation
    open Chessie.ErrorHandling

    type UserErrorMessage =
        | UserLoggedIn of string 
        | UsernameInvalid of string    
        
    let getErrorMessage e =
        match e with
        | UserLoggedIn msg -> msg
        | UsernameInvalid msg -> msg

    let login findUserFunction addUserFunction setLoggedInFunction username =
        let notEmpty = validateEmptyString (UsernameInvalid "Username is empty")
        let lengthValid = validateStringLength 1 20 (UsernameInvalid "Username is too long") 

        let validateUsername un = notEmpty un <* lengthValid un  
        
        let loginUser un =
            match findUserFunction un with
            | Some user -> 
                match user.IsLoggedIn with
                | true -> Bad [ (UserLoggedIn "User with that username is already logged in") ]
                | false -> 
                    setLoggedInFunction username |> ok
            | None ->
                { Id = 0; Username = un; IsLoggedIn = true } 
                |> addUserFunction
                |> ok

        username |> validateUsername >>= loginUser

    let logout findUserFunction setLoggedOutFunction username =
        match findUserFunction username with 
        | Some _ -> 
            setLoggedOutFunction username
            |> ok 
        | None -> Bad [ (UsernameInvalid "This user does not exist") ]


module UserServiceWithDataAccess =
    open PinetreeChat.DataAccess.UserRepository    
    let login = UserService.login getUser addUser setLoggedIn
    let logout = UserService.logout getUser setLoggedOut

            
        